[mono][eventpipe] Write primitive types and UTF-16 strings in little endian order

stefan-sf-ibm opened the pull request:
By convention of the [EventPipe file format](https://github.com/microsoft/perfview/blob/main/src/TraceEvent/EventPipe/EventPipeFormat.md) every primitive type is of little-endian order. Currently there is no distinction of the host endianness which means that while firing events primitive types are written in big-endian order on s390x. This, of course, renders the generated nettrace file unreadable e.g. by tools like perfview (especially if those files are read on little-endian machines). This patch fixes this by writing primitive types always in little-endian order which involves byte swaps on big-endian machines. In order to keep a good debug experience those byte swaps are done late.

The generated event functions by `genEventPipe.py` already utilize for primitive types specialized serialization functions as e.g. `write_buffer_uint32_t` which have been extended with byte swaps accordingly. In order to deal with events which contain arrays and (arrays of) structures the generated functions have been adapted as follows.

Arrays are iterated and each element is applied to a specialized function as e.g.

```c
uint32_t EventPipeWriteEventMethodDCEndILToNativeMap(
/* ... */
    for (uint32_t i = 0; i < CountOfMapEntries; ++i) {
        success &= write_buffer_uint32_t (ILOffsets[i], &buffer, &offset, &size, &fixedBuffer);
    }
    for (uint32_t i = 0; i < CountOfMapEntries; ++i) {
        success &= write_buffer_uint32_t (NativeOffsets[i], &buffer, &offset, &size, &fixedBuffer);
    }
```

Similar for structures each element is written by a specialized function as e.g.

```c
uint32_t EventPipeWriteEventMonoProfilerJitDone_V1(
/* ... */
    const uint8_t *valuePtr = Values;
    for (uint32_t i = 0; i < Count; ++i) {
        uint64_t value_uint64_t;
        uint8_t value_uint8_t;
        memcpy (&value_uint8_t, valuePtr, sizeof (value_uint8_t));
        valuePtr += sizeof (uint8_t);
        success &= write_buffer_uint8_t (value_uint8_t, &buffer, &offset, &size, &fixedBuffer);
        memcpy (&value_uint64_t, valuePtr, sizeof (value_uint64_t));
        valuePtr += sizeof (uint64_t);
        success &= write_buffer_uint64_t (value_uint64_t, &buffer, &offset, &size, &fixedBuffer);
    }
```

Following the idea of `genEventPipe.py` where for each primitive type `T` a specialized serialization function `write_buffer_T` exists this patch introduces the same for `block` and `fast serializer`. For the former new functions of the form `ep_write_buffer_T` have been introduced and for the latter calls to `ep_fast_serializer_write_buffer` are replaced by specialized functions as e.g. `ep_fast_serializer_write_uint32_t` where the byte swap happens.

lateralusX reviewed (commented):
Overall it looks great! Thank you for putting effort into this, extending EventPipe to support big endian architectures. Left some comments/questions/ideas in the PR.

josalem commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2776,7 +2776,13 @@ ep_rt_mono_sample_profiler_write_sampling_event_for_threads (
 			}
 			mono_thread_info_set_tid (&adapter, ep_rt_uint64_t_to_thread_id_t (data->thread_id));
-			ep_write_sample_profile_event (sampling_thread, sampling_event, &adapter, &data->stack_contents, (uint8_t *)&data->payload_data, sizeof (data->payload_data));
+			uint32_t payload_data =
+#if BIGENDIAN
+				__builtin_bswap32 (data->payload_data);
+#else
+				data->payload_data;
+#endif
+			ep_write_sample_profile_event (sampling_thread, sampling_event, &adapter, &data->stack_contents, (uint8_t *)&payload_data, sizeof (payload_data));
```
Rather than do these payload conversions here, it may be simpler to identify a big-endian arch at the reader end and handle it there. Doing it here will require you to find _every_ event, including user defined ones, with a payload and convert the contents to little-endian before it gets cast to a byte array. You'll either only have a few events with little-endian payloads, or you'll spend a lot more time converting everything and re-engineering all the code paths to swap bytes around.

noahfalk replied:
I think this is a question of scope. If @stefan-sf-ibm knows he won't care about EventSource generated events then the current approach handles it. @stefan-sf-ibm if you do care about those events then I could imagine a few ways to deal with it:
a. Do similar endian swapping loging in EventSource serializer and consider all events manually written via WriteEventCore as out of scope
b. Write some code that will parse event metadata to understand the schema of the payload and then endian swap all the fields (this handles the WriteEventCore case not handled in (a)
c. Make a breaking change to the file format allowing for big endian encodings and update format readers to accomodate it.

(b) adds more perf overhead, (c) probably creates a situation where tools based on sufficiently new TraceEvent support big endian data but older tools or non-TraceEvent based tools don't (I think we're up to at least three 3rd party implementations).

My inclination is to favor runtime complexity over file format complexity (favor the current PR approach or a/b options over the c option) assuming that big endian devices remain a tiny fraction of overall usage.

lateralusX replied:
I think that the current PR approach and the scope that it covers, payloads generated through https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/ClrEtwAll.man and emitted by Mono runtime, is a good start, I guess it mainly covers the initial TARGET_S390X needs, AFIK, that is the only big endian architecture currently running Mono out of dotnet/runtime (don't think CoreCLR supports any big endian platforms). If more support for custom payloads are needed in the future for big endian platforms, I believe we could revisit this, but current PR will at least make sure all Mono runtime emitted events are readable by existing tools.

stefan-sf-ibm replied:
Please see my comment regarding the overhead for little-endian architectures in the above [comment](https://github.com/dotnet/runtime/pull/68648#discussion_r868882196).

Ideally we would have for nettrace format something similar like in the Common Trace Format used by LTTng where one can define globally or even per integer (probably not important for us) the [byte order](https://diamon.org/ctf/#spec4.1.3). This would have the least overhead while taking a trace which is performance critical. My experience with EventPipe is limited to the last couple of weeks where I basically only got the implementation running. So I'm not sure how important all this is and how it manifests in practice like skipping events due to high load only because of byte swaps and its side-effects from the implementation like non-inlined function calls and so on. I'm also not sure how likely it is to change the current format and in which time frame something like that could happen. If there is anything I can do I would be happy to jump in. Anyhow, in order to gain some experience this PR might suite as a starter where all Mono events are handled. Though, in the long run I'm still kind of worried about user events.

@josalem the TraceEvent library is also on my todo list. A nettrace file created with and without this PR is not convertible on a big-endian machine. With this PR it is at least convertible on a little-endian machine. However, I just gave the sample provider a chance yesterday and traced an example application and the trace looks broken to me, i.e., the flamegraph contains no function names. It looks like as I have missed something which I need to investigate ... setting this PR to WIP.

lateralusX replied:
Make sure you get the rundown events into the .nettrace files and that they are correct since they are needed for the  symbolication of callstacks it tools, they have the mapping between method id and the function names/signatures. You can look at all events included in the nettrace file in perfview, and on some, it will also display the payload, but not for all.

josalem reviewed (commented):
I have a few thoughts on this:
* Echoing Johan's concerns about potential overhead for little-endian systems.
* The shared code paths, e.g., writing blocks, headers, metadata, etc. to the stream, should be little-endian for sure, but attempting to convert all the payloads seems like a Sisyphean task. It will probably be simpler to just allow payloads to be big endian and pay the price at the reader end to parse the payloads differently.
* To that end, we would need a mechanism for identifying to the reader that the incoming stream was big-endian payloads. This could be as simple as the reader accepting the magic `NETTRACE` bytes in little-endian or big-endian order and then adjusting its parsing accordingly.
* ~If we do things that way, CoreCLR would also getwork for big-endian archs.~ Let me validate CoreCLR support for big-endian archs before giving this feedback.

Lastly, out of curiosity, have you tried inspecting these big-endian files on a big-endian machine? Specifically, have you tried doing something like `dotnet-trace convert --fornat speedscope ...` and then opening _that_ file elsewhere?

stefan-sf-ibm commented on src/coreclr/scripts/genEventPipe.py:
```diff
@@ -283,9 +301,10 @@ def generateWriteEventBody(template, providerName, eventName, runtimeFlavor):
-                pack_list.append(
-                    "    success &= write_buffer((const uint8_t *)%s, %s, &buffer, &offset, &size, &fixedBuffer);" %
-                    (paramName, size))
+                t = winTypeToFixedWidthType(parameter.winType)
```
Thank you very much for the detailed review :)

My intention of having only one endian-independent implementation was that this will be better tested than having two obviously. Having a look at the assembly of a release build I see that for `EventPipeWriteEventMethodDCEndILToNativeMap` as well as `EventPipeWriteEventMonoProfilerJitDone_V1` all calls to `write_buffer_x` and `memcpy` are inlined (on x64 and s390x), however, this does not include the call to `write_buffer` from `eventpipehelpers.c` of course. Thus, in order to keep the overhead minimal and not introduce any potential performance regression I restored the old behavior for little-endian architectures.

Note, while having a further look at all `write_buffer_x` functions I changed `write_buffer_uintptr_t` slightly in case of `sizeof (uintptr_t) == 4`.

stefan-sf-ibm commented on src/coreclr/scripts/genEventPipe.py:
```diff
@@ -267,9 +278,16 @@ def generateWriteEventBody(template, providerName, eventName, runtimeFlavor):
-                pack_list.append(
-                    "    success &= write_buffer((const uint8_t *)%s, %s, &buffer, &offset, &size, &fixedBuffer);" %
-                    (paramName, size))
+                pack_list.append("    const uint8_t *valuePtr = %s;" % paramName)
```
See comment above.

stefan-sf-ibm commented on src/coreclr/scripts/genEventPipe.py:
```diff
@@ -851,6 +877,9 @@ def getMonoEventPipeImplFilePrefix():
     size_t *size,
     bool *fixed_buffer)
 {
+#if BIGENDIAN
```
I removed `EP_SWAPBYTES` and introduced `EP_UINT16_TO_LE` and similar in `ep.h`.

stefan-sf-ibm commented on src/native/eventpipe/ep-block.c:
```diff
@@ -909,12 +898,10 @@ ep_sequence_point_block_init (
 	const ep_timestamp_t timestamp = ep_sequence_point_get_timestamp (sequence_point);
-	memcpy (sequence_point_block->block.write_pointer, &timestamp, sizeof (timestamp));
-	sequence_point_block->block.write_pointer += sizeof (timestamp);
+	ep_write_buffer_int64_t (&sequence_point_block->block.write_pointer, timestamp);
```
Added function `ep_write_buffer_timestamp` which serves as a wrapper for `ep_write_buffer_int64_t`.

stefan-sf-ibm commented on src/native/eventpipe/ep-file.c:
```diff
@@ -93,15 +93,22 @@ file_fast_serialize_func (void *object, FastSerializer *fast_serializer)
-	ep_fast_serializer_write_buffer (fast_serializer, (const uint8_t *)&file->file_open_system_time, sizeof (file->file_open_system_time));
-	ep_fast_serializer_write_buffer (fast_serializer, (const uint8_t *)&file->file_open_timestamp, sizeof (file->file_open_timestamp));
-	ep_fast_serializer_write_buffer (fast_serializer, (const uint8_t *)&file->timestamp_frequency, sizeof (file->timestamp_frequency));
+	ep_fast_serializer_write_uint16_t (fast_serializer, file->file_open_system_time.year);
```
Done

stefan-sf-ibm commented on src/native/eventpipe/ep-file.c:
```diff
@@ -93,15 +93,22 @@ file_fast_serialize_func (void *object, FastSerializer *fast_serializer)
+	ep_fast_serializer_write_uint16_t (fast_serializer, file->file_open_system_time.second);
+	ep_fast_serializer_write_uint16_t (fast_serializer, file->file_open_system_time.milliseconds);
+	ep_fast_serializer_write_int64_t (fast_serializer, file->file_open_timestamp);
+	ep_fast_serializer_write_int64_t (fast_serializer, file->timestamp_frequency);
```
Just to be clear. I only used it for `file->file_open_timestamp` and left `file->timestamp_frequency` as is. Not sure whether an extra function is worthwhile for that or not?

lateralusX replied:
timestamp_frequency is typed as int64_t so that is fine to serialize as int64_t.

stefan-sf-ibm commented on src/native/eventpipe/ep.h:
```diff
@@ -288,6 +288,63 @@ ep_system_time_get (EventPipeSystemTime *system_time)
+ep_write_buffer_string_utf16_t (uint8_t **buf, const ep_char16_t *str, size_t len)
+{
+#if BIGENDIAN
+	ep_char16_t *buf_str = (ep_char16_t *)*buf;
```
I see three functions used for UTF-16 string handling:

```c
bool write_buffer_string_utf8_to_utf16_t (const ep_char8_t *value, uint8_t **buffer, size_t *offset, size_t *size, bool *fixed_buffer);
void metadata_generator_write_string_to_buffer (uint8_t *buffer, size_t buffer_len, size_t *offset, const ep_char16_t *value);
void ep_write_buffer_string_utf16_t (uint8_t **buf, const ep_char16_t *str, size_t len);
```

The first one does a string conversion which the latter two don't. Furthermore, all `metadata_generator_write_X_to_buffer` do not alter the pointer but change the offset in contrast to `ep_write_buffer_X` which alter the buffer pointer. Thus the API differs quite a bit. Despite that the body more or less only consists of `memcpy`. So I'm not sure how one implementation should look like. Did you have anything in mind?

lateralusX replied:
Not really, just saw that we did string conversions, but if there is nothing to reuse between the constructions, then you can keep it as is.

lateralusX commented on src/native/eventpipe/ep.h:
```diff
@@ -8,6 +8,22 @@
 #include "ep-stack-contents.h"
 #include "ep-rt.h"
 
+#if FEATURE_CORECLR
```
better to but this into each runtimes ep-rt.h header, that's where we keep runtime specializations.

stefan-sf-ibm commented:
### StackBlocks

The reason for missing all functions in the flame graph was that IPs of StackBlocks missed a byte swap. Fixed by adjusting `ep_stack_contents_append` and introducing `EP_UINTPTR_TO_LE`.

### UTF-16 strings

Some of the UTF-16 strings generated by `ep_rt_utf8_to_utf16_string` were properly converted in the end to UTF-16LE e.g. via `metadata_generator_write_string_to_buffer` or `ep_write_buffer_string_utf16_t` but some were not. For the payload (command line, os information, arch information) of the event `ProcessInfo` of the `Microsoft-DotNETCore-EventPipe` provider those were still [written in UTF-16BE](https://github.com/dotnet/runtime/blob/951df04d97405ff6dd13e708873549fb9acf9a6c/src/native/eventpipe/ep-event-source.c#L216). In order to fix this I went with doing the byte swap during conversion from UTF-8 to UTF-16 and introduced `g_utf8_to_utf16le`. This should also give a better performance since a UTF-16 string doesn't need to be reiterated. Furthermore, this should ensure that all UTF-16 strings which are allocated in the eventpipe namespace are little endian so we have a consistent layout of UTF-16 strings.

The converse, i.e., conversion from UTF-16 to UTF-8, needs some adjustment, too. Function `ep_rt_utf16_to_utf8_string` is used in two different contexts ([ep-rt-mono.c](https://github.com/dotnet/runtime/blob/f4d03106f43a95d731813d045938955d54345e32/src/mono/mono/eventpipe/ep-rt-mono.c#L3485), [ds-eventpipe-protocol.c](https://github.com/dotnet/runtime/blob/f4d03106f43a95d731813d045938955d54345e32/src/native/eventpipe/ds-eventpipe-protocol.c#L185)) where one time the input UTF-16 string is in host-endian order and the other time in little-endian order. In order to deal with both I added function `ep_rt_utf16le_to_utf8_string`. Thus following Mono's implementation which uses `utf16`, `utf16be`, `utf16le` for host-endian, big-endian, little-endian, respectively.

Currently the naming scheme differs for UTF-8 to UTF-16 conversion where `ep_rt_utf8_to_utf16_string` should actually be named `ep_rt_utf8_to_utf16le_string`. If you are fine I would like to rename this function.

### DS_VAL{16,32,64}

I'm wondering whether it is fine to replace all `DS_VAL{16,32,64}` macros defined in [ds-types.h](https://github.com/dotnet/runtime/blob/b1f6d216ccb15bf800d3d0c18ef152b67e8363a3/src/native/eventpipe/ds-types.h#L115) by `EP_UINT{16,32,64}_TO_LE`? Would imply that `ep-rt.h` must be included in the diagnostics namespace. Not sure whether this is acceptable or not.

At least the current implementation is missing casts. For example, in [here](https://github.com/dotnet/runtime/blob/9d0644972eca8caa05f8faaf9ca4b063c7fa75cc/src/native/eventpipe/ds-protocol.c#L105) the parameter is of type `uint32_t` which is not upcasted to an `uint64_t` by `DS_VAL64` which results in a bad shift. Thus, if `EP_UINT{16,32,64}_TO_LE` cannot be used I would like to replace those with `GUINT{16,32,64}_TO_LE`.

I will do some more testing in the following days.

Last but not least I just have to say that EventTrace is an awesome library which helped a lot in order to debug the binary trace file :)

stefan-sf-ibm commented:
As previously mentioned I removed the macros `DS_VAL{16,32,64}` and streamlined them with `EP_{U,}INT{16,32,64}_TO_LE` into proper C functions `ep_rt_val_uint32_t` as for example. I went for C functions instead of macros in order to rule out that arguments are evaluated multiple times like done [here](https://github.com/dotnet/runtime/blob/f4d03106f43a95d731813d045938955d54345e32/src/native/eventpipe/ds-protocol.c#L105)

```c
uint64_t pid = DS_VAL64 (ep_rt_current_process_get_id ());
```

Using attribute `EP_ALWAYS_INLINE` should ensure that there is no performance difference between a function and macro.

Beside that I fixed a couple of missing conversions in diagnostics. Since all my local tests are successful, now, removing draft flag.

stefan-sf-ibm committed:
[mono] Add endian-aware UTF-16 string conversion functions

stefan-sf-ibm committed:
[mono][eventpipe] Add big-endian support

stefan-sf-ibm commented:
While rebasing in order to fix some merge conflicts new events where introduced which require string handling as e.g. `EventPipeWriteEventBulkType`. Added support for those, too.

stefan-sf-ibm commented:
@lateralusX I did a couple of further tests and so far they all passed. Can we merge this?

lateralusX commented:
> @lateralusX I did a couple of further tests and so far they all passed. Can we merge this?

@stefan-sf-ibm I looked through the changes again I believe they look fine.

@noahfalk Any objections merging this? The initial concerns have been addressed (so little endian systems should have close to zero overhead) and it will make sure native runtime events works on big endian systems enable use of perfview on nettrace files captured on IBM S390X. It will not do anything with managed events since there is currently not hooked into the serializer mechanism of those events, but based on above discussion, at least getting the native runtime events working with existing tooling adds value for big endian platforms worth doing the extension.

lateralusX commented:
> Would imply that `ep-rt.h` must be included in the diagnostics namespace. Not sure whether this is acceptable or not.

That is OK, DS namespace already have relationship -> EP headers, what we should avoid is the other way around.

uweigand commented:
Merging this will also fix the `ThreadPoolMinMaxThreadsEventTest` regression on s390x discussed in https://github.com/dotnet/runtime/issues/71173.

noahfalk commented:
No objections from me, sorry to keep you waiting I was out sick the last couple days.
cc @davmason

davmason commented:
No objections from me either

lateralusX commented:
Failures in runtime-extra-platforms looks unrelated.

lateralusX commented:
/azp run runtime

stefan-sf-ibm commented:
I could also try a rebase if that is fine for you. I remember using a two weeks old mainline version I ran into some weird problems which seem to be fixed meanwhile.

lateralusX commented:
@stefan-sf-ibm, yes please rebase and we can see if we get better pass rate on runtime-extra-platforms as well.

lateralusX commented:
/azp run runtime-extra-platforms

lateralusX reviewed (approved):
LGTM! Thanks for adding support for big endian platform in EventPipe native runtime events!
